library/src/scala/IArray.scala

```diff
@@ -1,288 +1,316 @@
-    extension [T](arr: IArray[T]) def startsWith[U >: T: ClassTag](that: IArray[U], offset: Int = 0): Boolean =
-      genericArrayOps(arr).startsWith(that)
+object IArray:
+
+  /** The selection operation on an immutable array.
+    *
+    *  @param arr the immutable array
+    *  @param n   the index of the element to select
+    *  @return    the element of the array at the given index
+    */
+  extension (arr: IArray[Byte]) def apply(n: Int): Byte = arr.asInstanceOf[Array[Byte]].apply(n)
+  extension (arr: IArray[Short]) def apply(n: Int): Short = arr.asInstanceOf[Array[Short]].apply(n)
+  extension (arr: IArray[Char]) def apply(n: Int): Char = arr.asInstanceOf[Array[Char]].apply(n)
+  extension (arr: IArray[Int]) def apply(n: Int): Int = arr.asInstanceOf[Array[Int]].apply(n)
+  extension (arr: IArray[Long]) def apply(n: Int): Long = arr.asInstanceOf[Array[Long]].apply(n)
+  extension (arr: IArray[Float]) def apply(n: Int): Float = arr.asInstanceOf[Array[Float]].apply(n)
+  extension (arr: IArray[Double]) def apply(n: Int): Double = arr.asInstanceOf[Array[Double]].apply(n)
+  extension [T <: Object](arr: IArray[T]) def apply (n: Int): T = arr.asInstanceOf[Array[T]].apply(n)
+  extension [T](arr: IArray[T]) def apply (n: Int): T = arr.asInstanceOf[Array[T]].apply(n)
+
+  /** The number of elements in an immutable array
+    *  @param arr  the immutable array
+    */
+  extension (arr: IArray[Byte]) def length: Int = arr.asInstanceOf[Array[Byte]].length
+  extension (arr: IArray[Short]) def length: Int = arr.asInstanceOf[Array[Short]].length
+  extension (arr: IArray[Char]) def length: Int = arr.asInstanceOf[Array[Char]].length
+  extension (arr: IArray[Int]) def length: Int = arr.asInstanceOf[Array[Int]].length
+  extension (arr: IArray[Long]) def length: Int = arr.asInstanceOf[Array[Long]].length
+  extension (arr: IArray[Float]) def length: Int = arr.asInstanceOf[Array[Float]].length
+  extension (arr: IArray[Double]) def length: Int = arr.asInstanceOf[Array[Double]].length
+  extension (arr: IArray[Object]) def length: Int = arr.asInstanceOf[Array[Object]].length
+  extension [T](arr: IArray[T]) def length: Int = arr.asInstanceOf[Array[T]].length
+
+  /** Returns this array concatenated with the given array. */
+  extension [T](arr: IArray[T]) def ++ [U >: T: ClassTag](that: IArray[U]): IArray[U] =
+    genericArrayOps(arr) ++ that
+
+  /** Tests whether this array contains a given value as an element. */
+  extension [T](arr: IArray[T]) def contains(elem: T): Boolean =
+    // `genericArrayOps(arr).contains(elem)` does not work because `elem` does not have type `arr.T`
+    // but we can use `exists` instead, which is how `ArrayOps#contains` itself is implemented:
+    genericArrayOps(arr).exists(_ == elem)
+
+  /** Copy elements of this array to another array. */
+  extension [T](arr: IArray[T]) def copyToArray[U >: T](xs: Array[U]): Int =
+    genericArrayOps(arr).copyToArray(xs)
```

Comment:
how about collective style?

Reply:
Could work. This code comes from before we had collective extensions. That would not change the API or generated code though.

Reply:
Looking at the operations we have, we should probably define extensions in groups with all overloads of the same method. All `apply` in one collective extension, all `length` in another collective extension and so on.

Reply:
Actually, we can only do a collective extension on arrays that have the same type. That will not work if we want to keep specialized versions of the same method together.